File: TidbitsTrivia/Core/Networking/FirebaseRTDB.swift
import Foundation

/// Firebase Realtime Database over its REST + streaming API — the Apple twin of
/// `js/firebase.js` and `FirebaseNet.kt`, but with NO third-party SDK (URLSession
/// only, per the Apple no-packages rule). Anonymous auth via the Identity Toolkit
/// REST endpoint; reads/writes via the RTDB `.json` REST API; live updates via
/// Server-Sent Events (`text/event-stream`).
///
/// Shared by the Mac host (publishing a Tidbits Live room) and the iOS join
/// surface. Same project + non-secret config as the web/Android clients, so all
/// four platforms meet in ONE room (js/firebase-config.js). Access is gated by
/// Security Rules + the anon-auth requirement, not by hiding the config.
///
/// Verified end to end against the live project with curl (anon sign-up → PUT →
/// GET → PATCH → roster read → delete) before this client was written.
actor FirebaseRTDB {
    struct Config: Sendable {
        let apiKey: String
        let databaseURL: String   // https://<project>-default-rtdb.firebaseio.com
    }

    /// The committed non-secret web config (same project as js/firebase-config.js).
    static let shared = FirebaseRTDB(config: .init(
        apiKey: "AIzaSyCns8iba6zVqkddEUY_gqoc4eVxz-3BGaA",
        databaseURL: "https://tidbits-trivia-f2ddb-default-rtdb.firebaseio.com"))

    let config: Config
    private let session: URLSession
    private var idToken: String?
    private var refreshToken: String?
    private var expiry: Date = .distantPast
    private(set) var uid: String?

    init(config: Config, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    enum RTDBError: Error, Equatable { case http(Int), noToken, badResponse, alreadyLinked }

    /// Outcome of a federated sign-in: the provider account's uid + its verified email
    /// (identity keys the shared profile by the email, not the uid).
    struct FederatedResult: Sendable { let uid: String; let email: String? }

    // MARK: - Auth (anonymous)

    /// Ensure a live anonymous session; returns the uid. Signs up on first use,
    /// refreshes when the hour-long token nears expiry.
    @discardableResult
    func ensureAuth() async throws -> String {
        if let uid, Date() < expiry.addingTimeInterval(-300) { return uid }
        if refreshToken == nil { refreshToken = Keychain.get(Self.refreshKey) }   // restore persisted session
        if refreshToken != nil {
            do { try await refreshSession() }
            catch { refreshToken = nil; Keychain.delete(Self.refreshKey); try await signUpAnonymous() }
        } else {
            try await signUpAnonymous()
        }
        guard let uid else { throw RTDBError.noToken }
        return uid
    }

    private struct SignUpResponse: Decodable {
        let idToken: String; let refreshToken: String; let expiresIn: String; let localId: String
    }
    private func signUpAnonymous() async throws {
        var req = URLRequest(url: URL(string: "https://identitytoolkit.googleapis.com/v1/accounts:signUp?key=\(config.apiKey)")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: ["returnSecureToken": true])
        let (data, resp) = try await session.data(for: req)
        try Self.check(resp)
        let r = try JSONDecoder().decode(SignUpResponse.self, from: data)
        apply(idToken: r.idToken, refreshToken: r.refreshToken, expiresIn: r.expiresIn, uid: r.localId)
    }

    private struct RefreshResponse: Decodable {
        let id_token: String; let refresh_token: String; let expires_in: String; let user_id: String
    }
    private func refreshSession() async throws {
        guard let refreshToken else { try await signUpAnonymous(); return }
        var req = URLRequest(url: URL(string: "https://securetoken.googleapis.com/v1/token?key=\(config.apiKey)")!)
        req.httpMethod = "POST"
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        req.httpBody = "grant_type=refresh_token&refresh_token=\(refreshToken)".data(using: .utf8)
        let (data, resp) = try await session.data(for: req)
        try Self.check(resp)
        let r = try JSONDecoder().decode(RefreshResponse.self, from: data)
        apply(idToken: r.id_token, refreshToken: r.refresh_token, expiresIn: r.expires_in, uid: r.user_id)
    }

    private struct IdpResponse: Decodable {
        let idToken: String; let refreshToken: String; let expiresIn: String; let localId: String
        let email: String?
    }

    /// Sign in with Apple → the Apple Firebase account (its own uid; allowDuplicateEmails
    /// lets it coexist with a Google account for the same email). Identity keys the shared
    /// profile by the verified email so both providers land on the same records.
    func signInWithApple(identityToken: String, rawNonce: String) async throws -> FederatedResult {
        let post = "id_token=\(identityToken)&providerId=apple.com&nonce=\(rawNonce)"
        let r = try await signInWithIdp(postBody: post, linkTo: nil)
        apply(idToken: r.idToken, refreshToken: r.refreshToken, expiresIn: r.expiresIn, uid: r.localId)
        return FederatedResult(uid: r.localId, email: r.email)
    }

    /// The verified email of the current federated session, decoded from the Firebase ID
    /// token payload (nil for anonymous). Lets the profile re-key by email after relaunch.
    func currentEmail() -> String? {
        guard let token = idToken else { return nil }
        let parts = token.split(separator: ".")
        guard parts.count == 3 else { return nil }
        var b64 = String(parts[1]).replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while b64.count % 4 != 0 { b64 += "=" }
        guard let data = Data(base64Encoded: b64),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return json["email"] as? String
    }

    /// Sign out of the federated account and return to a FRESH anonymous session (new
    /// uid). The account's records stay in players/{accountUid}; signing back in restores.
    func signOut() async throws -> String {
        Keychain.delete(Self.refreshKey)
        idToken = nil; refreshToken = nil; uid = nil; expiry = .distantPast
        try await signUpAnonymous()
        return uid ?? ""
    }

    private func signInWithIdp(postBody: String, linkTo idToken: String?) async throws -> IdpResponse {
        var body: [String: Any] = ["postBody": postBody, "requestUri": "http://localhost",
                                   "returnSecureToken": true, "returnIdpCredential": true]
        if let idToken { body["idToken"] = idToken }
        var req = URLRequest(url: URL(string: "https://identitytoolkit.googleapis.com/v1/accounts:signInWithIdp?key=\(config.apiKey)")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await session.data(for: req)
        if let http = resp as? HTTPURLResponse, http.statusCode >= 400 {
            let msg = String(data: data, encoding: .utf8) ?? ""
            if msg.contains("FEDERATED_USER_ID_ALREADY_LINKED") || msg.contains("EMAIL_EXISTS") { throw RTDBError.alreadyLinked }
            throw RTDBError.http(http.statusCode)
        }
        return try JSONDecoder().decode(IdpResponse.self, from: data)
    }

    /// Keychain key for the anonymous refresh token — persisting it keeps the `uid`
    /// (hence the portable player profile) stable across relaunches.
    static let refreshKey = "tidbits.fb.anonRefresh"

    private func apply(idToken: String, refreshToken: String, expiresIn: String, uid: String) {
        self.idToken = idToken
        self.refreshToken = refreshToken
        self.uid = uid
        self.expiry = Date().addingTimeInterval(TimeInterval(expiresIn) ?? 3600)
        Keychain.set(refreshToken, for: Self.refreshKey)
    }

    private func validToken() async throws -> String {
        try await ensureAuth()
        guard let idToken else { throw RTDBError.noToken }
        return idToken
    }

    // MARK: - REST read/write

    /// Overwrite the value at `path` (RTDB PUT).
    func put<T: Encodable>(_ path: String, _ value: T) async throws {
        try await write(path, value, method: "PUT")
    }
    /// Merge children at `path` (RTDB PATCH).
    func patch<T: Encodable>(_ path: String, _ value: T) async throws {
        try await write(path, value, method: "PATCH")
    }
    func delete(_ path: String) async throws {
        let token = try await validToken()
        var req = URLRequest(url: restURL(path, token: token))
        req.httpMethod = "DELETE"
        let (_, resp) = try await session.data(for: req)
        try Self.check(resp)
    }
    /// True if a value exists (non-null) at `path`. A cheap presence check that
    /// decodes nothing — so it can probe a node keyed by a @MainActor-isolated
    /// Codable type (e.g. `live/{code}/meta`) without tripping Swift 6's
    /// isolated-conformance rule inside this actor. Used by the unified "Join a
    /// game" front to tell a hosted Tidbits Live room from a LAN Trivia Night.
    func exists(_ path: String) async throws -> Bool {
        let token = try await validToken()
        let (data, resp) = try await session.data(from: restURL(path, token: token))
        try Self.check(resp)
        guard let s = String(data: data, encoding: .utf8) else { return false }
        return !data.isEmpty && s != "null"
    }

    /// Read + decode the value at `path` (nil if the node is absent → JSON null).
    func get<T: Decodable>(_ path: String, as type: T.Type) async throws -> T? {
        let token = try await validToken()
        let (data, resp) = try await session.data(from: restURL(path, token: token))
        try Self.check(resp)
        if data.isEmpty || String(data: data, encoding: .utf8) == "null" { return nil }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func write<T: Encodable>(_ path: String, _ value: T, method: String) async throws {
        try await writeJSON(path, JSONEncoder().encode(value), method: method)
    }

    /// Write pre-encoded JSON. Callers on another actor (e.g. the @MainActor host)
    /// encode their own value first, so a MainActor-isolated Codable conformance
    /// never has to run inside this actor (Swift 6 isolated-conformances).
    func putJSON(_ path: String, _ json: Data) async throws { try await writeJSON(path, json, method: "PUT") }
    func patchJSON(_ path: String, _ json: Data) async throws { try await writeJSON(path, json, method: "PATCH") }

    private func writeJSON(_ path: String, _ json: Data, method: String) async throws {
        let token = try await validToken()
        var req = URLRequest(url: restURL(path, token: token))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = json
        let (_, resp) = try await session.data(for: req)
        try Self.check(resp)
    }

    // MARK: - Live streaming (Server-Sent Events)

    struct StreamEvent: Sendable {
        let event: String     // "put" | "patch" | "keep-alive" | "cancel" | "auth_revoked"
        let path: String      // relative to the streamed node, e.g. "/" or "/teams/uid"
        let dataJSON: Data?    // the SSE `data` value re-encoded as JSON (nil when null)
    }

    /// Stream live updates at `path` as SSE. The stream cancels its underlying
    /// request when the consumer's task is cancelled.
    func stream(_ path: String) async throws -> AsyncThrowingStream<StreamEvent, Error> {
        let token = try await validToken()
        var req = URLRequest(url: restURL(path, token: token))
        req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 3600
        return Self.makeStream(req: req, session: session)
    }

    /// nonisolated so the AsyncThrowingStream build closure captures only Sendable
    /// values (URLRequest, URLSession) — never actor-isolated state (Swift 6).
    nonisolated static func makeStream(req: URLRequest, session: URLSession) -> AsyncThrowingStream<StreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let (bytes, resp) = try await session.bytes(for: req)
                    try check(resp)
                    var currentEvent = ""
                    for try await line in bytes.lines {
                        if line.hasPrefix("event: ") {
                            currentEvent = String(line.dropFirst(7))
                        } else if line.hasPrefix("data: ") {
                            if let ev = parse(event: currentEvent, payload: String(line.dropFirst(6))) { continuation.yield(ev) }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Parse an RTDB SSE frame: `data: {"path":"/x","data":<value>}`.
    nonisolated static func parse(event: String, payload: String) -> StreamEvent? {
        guard event != "keep-alive" else { return nil }
        guard let d = payload.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: d) as? [String: Any] else {
            return StreamEvent(event: event, path: "/", dataJSON: nil)
        }
        let path = obj["path"] as? String ?? "/"
        var dataJSON: Data?
        if let value = obj["data"], !(value is NSNull) {
            dataJSON = try? JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed])
        }
        return StreamEvent(event: event, path: path, dataJSON: dataJSON)
    }

    // MARK: - Helpers

    private func restURL(_ path: String, token: String) -> URL {
        let clean = path.hasPrefix("/") ? String(path.dropFirst()) : path
        return URL(string: "\(config.databaseURL)/\(clean).json?auth=\(token)")!
    }

    nonisolated static func check(_ resp: URLResponse) throws {
        guard let http = resp as? HTTPURLResponse else { throw RTDBError.badResponse }
        guard (200...299).contains(http.statusCode) else { throw RTDBError.http(http.statusCode) }
    }

    /// A short human-shareable room code (Crockford-ish, no confusable chars) —
    /// the same alphabet as js/firebase.js so codes read identically everywhere.
    static func newRoomCode() -> String {
        let alphabet = Array("ABCDEFGHJKLMNPQRSTUVWXYZ23456789")
        return String((0..<4).map { _ in alphabet.randomElement()! })
    }
}

File: windows/Tidbits.Core/Networking/FirebaseRtdb.cs
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Tidbits.Core.Networking;

/// Firebase Realtime Database over its REST + SSE API — the C# twin of js/firebase.js
/// and the Swift FirebaseRTDB, no SDK (HttpClient only). Anonymous auth via Identity
/// Toolkit; reads/writes via the RTDB `.json` REST API; live updates via Server-Sent
/// Events. Same project + non-secret config as web/Android, so all clients meet in ONE
/// room; access is gated by Security Rules + anon-auth, not by hiding the config.
public sealed class FirebaseRtdb
{
    public sealed record Config(string ApiKey, string DatabaseUrl);

    /// The committed non-secret web config (same project as js/firebase-config.js).
    public static readonly Config DefaultConfig = new(
        "AIzaSyCns8iba6zVqkddEUY_gqoc4eVxz-3BGaA",
        "https://tidbits-trivia-f2ddb-default-rtdb.firebaseio.com");

    public const string RefreshKey = "tidbits.fb.anonRefresh";

    private readonly Config _config;
    private readonly HttpClient _http;
    private readonly ITokenStore _tokens;
    private readonly SemaphoreSlim _authLock = new(1, 1);

    private string? _idToken;
    private string? _refreshToken;
    private DateTime _expiry = DateTime.MinValue;
    public string? Uid { get; private set; }

    public FirebaseRtdb(Config? config = null, HttpClient? http = null, ITokenStore? tokens = null)
    {
        _config = config ?? DefaultConfig;
        _http = http ?? new HttpClient();
        _tokens = tokens ?? new FileTokenStore();
    }

    public sealed record FederatedResult(string Uid, string? Email, string? DisplayName);

    public class RtdbException(int status) : Exception($"RTDB HTTP {status}") { public int Status { get; } = status; }
    public sealed class AlreadyLinkedException() : Exception("federated user already linked");

    // MARK: Auth (anonymous)

    public async Task<string> EnsureAuth()
    {
        await _authLock.WaitAsync();
        try
        {
            if (Uid is not null && DateTime.UtcNow < _expiry.AddSeconds(-300)) return Uid;
            _refreshToken ??= _tokens.Get(RefreshKey);
            if (_refreshToken is not null)
            {
                try { await RefreshSession(); }
                catch { _refreshToken = null; _tokens.Delete(RefreshKey); await SignUpAnonymous(); }
            }
            else await SignUpAnonymous();
            return Uid ?? throw new RtdbException(401);
        }
        finally { _authLock.Release(); }
    }

    private async Task SignUpAnonymous()
    {
        using var resp = await _http.PostAsync(
            $"https://identitytoolkit.googleapis.com/v1/accounts:signUp?key={_config.ApiKey}",
            new StringContent("{\"returnSecureToken\":true}", Encoding.UTF8, "application/json"));
        Check(resp);
        var r = JsonSerializer.Deserialize<SignUpResponse>(await resp.Content.ReadAsStringAsync())!;
        Apply(r.IdToken, r.RefreshToken, r.ExpiresIn, r.LocalId);
    }

    private async Task RefreshSession()
    {
        if (_refreshToken is null) { await SignUpAnonymous(); return; }
        using var resp = await _http.PostAsync(
            $"https://securetoken.googleapis.com/v1/token?key={_config.ApiKey}",
            new StringContent($"grant_type=refresh_token&refresh_token={_refreshToken}",
                Encoding.UTF8, "application/x-www-form-urlencoded"));
        Check(resp);
        var r = JsonSerializer.Deserialize<RefreshResponse>(await resp.Content.ReadAsStringAsync())!;
        Apply(r.IdToken, r.RefreshToken, r.ExpiresIn, r.UserId);
    }

    public async Task<FederatedResult> SignInWithApple(string identityToken, string rawNonce)
    {
        var post = $"id_token={identityToken}&providerId=apple.com&nonce={rawNonce}";
        var r = await SignInWithIdp(post, null);
        Apply(r.IdToken, r.RefreshToken, r.ExpiresIn, r.LocalId);
        return new FederatedResult(r.LocalId, r.Email, r.DisplayName);
    }

    /// Google sign-in. `idToken` comes from the desktop OAuth loopback flow
    /// (`GoogleOAuth`); Firebase accepts it because the OAuth client lives in the same
    /// Google Cloud project. Same shape as SignInWithApple so both converge on the
    /// email-keyed profile.
    public async Task<FederatedResult> SignInWithGoogle(string idToken)
    {
        var post = $"id_token={Uri.EscapeDataString(idToken)}&providerId=google.com";
        var r = await SignInWithIdp(post, null);
        Apply(r.IdToken, r.RefreshToken, r.ExpiresIn, r.LocalId);
        return new FederatedResult(r.LocalId, r.Email, r.DisplayName);
    }

    public string? CurrentEmail() => _idToken is null ? null : EmailFromJwt(_idToken);

    /// True when the signed-in token carries a VERIFIED email. The RTDB rules require
    /// `auth.token.email_verified === true` to write the email-keyed profile, so the UI
    /// must not promise a synced identity when this is false.
    public bool CurrentEmailVerified() => _idToken is not null && EmailVerifiedFromJwt(_idToken);

    /// Decode the `email_verified` claim. Mirrors the rule the RTDB enforces.
    public static bool EmailVerifiedFromJwt(string token)
    {
        var parts = token.Split('.');
        if (parts.Length != 3) return false;
        var b64 = parts[1].Replace('-', '+').Replace('_', '/');
        while (b64.Length % 4 != 0) b64 += "=";
        try
        {
            using var doc = JsonDocument.Parse(Convert.FromBase64String(b64));
            if (!doc.RootElement.TryGetProperty("email_verified", out var v)) return false;
            return v.ValueKind == JsonValueKind.True ||
                   (v.ValueKind == JsonValueKind.String && v.GetString() == "true");
        }
        catch { return false; }
    }

    /// Decode the `email` claim from a JWT payload (Firebase ID token or Apple identity token).
    public static string? EmailFromJwt(string token)
    {
        var parts = token.Split('.');
        if (parts.Length != 3) return null;
        var b64 = parts[1].Replace('-', '+').Replace('_', '/');
        while (b64.Length % 4 != 0) b64 += "=";
        try
        {
            using var doc = JsonDocument.Parse(Convert.FromBase64String(b64));
            if (doc.RootElement.TryGetProperty("email", out var e) && e.ValueKind == JsonValueKind.String)
            {
                var email = e.GetString();
                return string.IsNullOrEmpty(email) ? null : email;
            }
        }
        catch { }
        return null;
    }

    public async Task<string> SignOut()
    {
        _tokens.Delete(RefreshKey);
        _idToken = null; _refreshToken = null; Uid = null; _expiry = DateTime.MinValue;
        await SignUpAnonymous();
        return Uid ?? "";
    }

    /// Delete the CREDENTIAL itself (Identity Toolkit `accounts:delete`), then start a fresh
    /// anonymous session so the app still works afterwards. Returns the new uid.
    ///
    /// Twin of Swift `FirebaseRTDB.deleteAccount()`. This is the authoritative step of account
    /// deletion — the RTDB node deletes around it are best-effort, but if this throws the
    /// account still exists and the caller MUST report failure rather than claim success.
    public async Task<string> DeleteAccount()
    {
        await EnsureAuth();
        var body = JsonSerializer.Serialize(new Dictionary<string, string> { ["idToken"] = _idToken ?? "" });
        using var resp = await _http.PostAsync(
            $"https://identitytoolkit.googleapis.com/v1/accounts:delete?key={_config.ApiKey}",
            new StringContent(body, Encoding.UTF8, "application/json"));
        Check(resp);
        _tokens.Delete(RefreshKey);
        _idToken = null; _refreshToken = null; Uid = null; _expiry = DateTime.MinValue;
        await SignUpAnonymous();
        return Uid ?? "";
    }

    private async Task<IdpResponse> SignInWithIdp(string postBody, string? linkTo)
    {
        var body = new Dictionary<string, object>
        {
            ["postBody"] = postBody, ["requestUri"] = "http://localhost",
            ["returnSecureToken"] = true, ["returnIdpCredential"] = true,
        };
        if (linkTo is not null) body["idToken"] = linkTo;
        using var resp = await _http.PostAsync(
            $"https://identitytoolkit.googleapis.com/v1/accounts:signInWithIdp?key={_config.ApiKey}",
            new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json"));
        var text = await resp.Content.ReadAsStringAsync();
        if ((int)resp.StatusCode >= 400)
        {
            if (text.Contains("FEDERATED_USER_ID_ALREADY_LINKED") || text.Contains("EMAIL_EXISTS"))
                throw new AlreadyLinkedException();
            throw new RtdbException((int)resp.StatusCode);
        }
        return JsonSerializer.Deserialize<IdpResponse>(text)!;
    }

    private void Apply(string idToken, string refreshToken, string expiresIn, string uid)
    {
        _idToken = idToken; _refreshToken = refreshToken; Uid = uid;
        _expiry = DateTime.UtcNow.AddSeconds(double.TryParse(expiresIn, out var s) ? s : 3600);
        _tokens.Set(RefreshKey, refreshToken);
    }

    private async Task<string> ValidToken()
    {
        await EnsureAuth();
        return _idToken ?? throw new RtdbException(401);
    }

    // MARK: REST read/write

    public Task Put<T>(string path, T value) => WriteJson(path, JsonSerializer.SerializeToUtf8Bytes(value, Wire.Json), HttpMethod.Put);
    public Task Patch<T>(string path, T value) => WriteJson(path, JsonSerializer.SerializeToUtf8Bytes(value, Wire.Json), HttpMethod.Patch);
    public Task PutJson(string path, byte[] json) => WriteJson(path, json, HttpMethod.Put);
    public Task PatchJson(string path, byte[] json) => WriteJson(path, json, HttpMethod.Patch);

    public async Task Delete(string path)
    {
        var token = await ValidToken();
        using var resp = await _http.DeleteAsync(RestUrl(path, token));
        Check(resp);
    }

    /// True if a value exists (non-null) at `path`.
    public async Task<bool> Exists(string path)
    {
        var token = await ValidToken();
        using var resp = await _http.GetAsync(RestUrl(path, token));
        Check(resp);
        var s = await resp.Content.ReadAsStringAsync();
        return s.Length > 0 && s != "null";
    }

    /// Read + deserialize the value at `path` (default(T) if absent → JSON null).
    public async Task<T?> Get<T>(string path)
    {
        var token = await ValidToken();
        using var resp = await _http.GetAsync(RestUrl(path, token));
        Check(resp);
        var s = await resp.Content.ReadAsStringAsync();
        if (s.Length == 0 || s == "null") return default;
        return JsonSerializer.Deserialize<T>(s, Wire.Json);
    }

    /// Read the RAW JSON at `path` (null if absent). A saved quiz decodes itself from
    /// the frozen contract bytes (docs/QUIZ-CONTRACT.md), so routing it through a
    /// typed Deserialize would create a SECOND representation of a format that
    /// already has four implementations.
    public async Task<string?> GetJson(string path)
    {
        var token = await ValidToken();
        using var resp = await _http.GetAsync(RestUrl(path, token));
        Check(resp);
        var s = await resp.Content.ReadAsStringAsync();
        return (s.Length == 0 || s == "null") ? null : s;
    }

    /// Write RAW JSON to `path` — the mirror of GetJson, same reasoning.
    public async Task PutJson(string path, string json)
    {
        var token = await ValidToken();
        using var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        using var resp = await _http.PutAsync(RestUrl(path, token), content);
        Check(resp);
    }

    /// Read a value together with its RTDB ETag, for a compare-and-set transaction
    /// (optimistic concurrency — the matchmaking queue claim, 2.21).
    public async Task<(T? Value, string ETag)> GetWithEtag<T>(string path)
    {
        var token = await ValidToken();
        using var req = new HttpRequestMessage(HttpMethod.Get, RestUrl(path, token));
        req.Headers.TryAddWithoutValidation("X-Firebase-ETag", "true");
        using var resp = await _http.SendAsync(req);
        Check(resp);
        var etag = resp.Headers.ETag?.Tag ?? "";
        var s = await resp.Content.ReadAsStringAsync();
        var value = (s.Length == 0 || s == "null") ? default : JsonSerializer.Deserialize<T>(s, Wire.Json);
        return (value, etag);
    }

    /// Conditional PUT — writes only if the value at `path` still matches `etag`.
    /// Returns false on 412 Precondition Failed (someone else won the race). Pass a
    /// null value to conditionally CLEAR the node (RTDB deletes on PUT null).
    public async Task<bool> CasPut<T>(string path, T? value, string etag)
    {
        var token = await ValidToken();
        var json = JsonSerializer.SerializeToUtf8Bytes(value, Wire.Json);
        using var req = new HttpRequestMessage(HttpMethod.Put, RestUrl(path, token))
        {
            Content = new ByteArrayContent(json),
        };
        req.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/json");
        req.Headers.TryAddWithoutValidation("if-match", etag);
        using var resp = await _http.SendAsync(req);
        if (resp.StatusCode == System.Net.HttpStatusCode.PreconditionFailed) return false;
        Check(resp);
        return true;
    }

    private async Task WriteJson(string path, byte[] json, HttpMethod method)
    {
        var token = await ValidToken();
        using var req = new HttpRequestMessage(method, RestUrl(path, token))
        {
            Content = new ByteArrayContent(json),
        };
        req.Content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/json");
        using var resp = await _http.SendAsync(req);
        Check(resp);
    }

    // MARK: Live streaming (SSE)

    public sealed record StreamEvent(string Event, string Path, string? DataJson);

    /// Stream live updates at `path` as SSE. Enumeration ends when the token is cancelled.
    public async IAsyncEnumerable<StreamEvent> Stream(string path, [EnumeratorCancellation] CancellationToken ct = default)
    {
        var token = await ValidToken();
        using var req = new HttpRequestMessage(HttpMethod.Get, RestUrl(path, token));
        req.Headers.Accept.ParseAdd("text/event-stream");
        using var resp = await _http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        resp.EnsureSuccessStatusCode();
        await using var stream = await resp.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream);

        string currentEvent = "";
        while (!ct.IsCancellationRequested)
        {
            var line = await reader.ReadLineAsync(ct);
            if (line is null) break;
            if (line.StartsWith("event: ")) currentEvent = line[7..];
            else if (line.StartsWith("data: "))
            {
                var ev = ParseEvent(currentEvent, line[6..]);
                if (ev is not null) yield return ev;
            }
        }
    }

    /// Parse an RTDB SSE frame: `data: {"path":"/x","data":<value>}`.
    public static StreamEvent? ParseEvent(string @event, string payload)
    {
        if (@event == "keep-alive") return null;
        try
        {
            using var doc = JsonDocument.Parse(payload);
            var root = doc.RootElement;
            var path = root.TryGetProperty("path", out var p) && p.ValueKind == JsonValueKind.String ? p.GetString()! : "/";
            string? dataJson = null;
            if (root.TryGetProperty("data", out var d) && d.ValueKind != JsonValueKind.Null)
                dataJson = d.GetRawText();
            return new StreamEvent(@event, path, dataJson);
        }
        catch
        {
            return new StreamEvent(@event, "/", null);
        }
    }

    // MARK: Helpers

    private string RestUrl(string path, string token)
    {
        var clean = path.StartsWith('/') ? path[1..] : path;
        return $"{_config.DatabaseUrl}/{clean}.json?auth={token}";
    }

    private static void Check(HttpResponseMessage resp)
    {
        if (!resp.IsSuccessStatusCode) throw new RtdbException((int)resp.StatusCode);
    }

    /// A short human-shareable room code — the SAME alphabet as js/firebase.js so codes
    /// read identically everywhere (Crockford-ish, no confusable chars).
    public static string NewRoomCode()
    {
        const string alphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
        var chars = new char[4];
        for (int i = 0; i < 4; i++) chars[i] = alphabet[Random.Shared.Next(alphabet.Length)];
        return new string(chars);
    }

    // Auth response DTOs
    private sealed record SignUpResponse(
        [property: JsonPropertyName("idToken")] string IdToken,
        [property: JsonPropertyName("refreshToken")] string RefreshToken,
        [property: JsonPropertyName("expiresIn")] string ExpiresIn,
        [property: JsonPropertyName("localId")] string LocalId);

    private sealed record RefreshResponse(
        [property: JsonPropertyName("id_token")] string IdToken,
        [property: JsonPropertyName("refresh_token")] string RefreshToken,
        [property: JsonPropertyName("expires_in")] string ExpiresIn,
        [property: JsonPropertyName("user_id")] string UserId);

    private sealed record IdpResponse(
        [property: JsonPropertyName("idToken")] string IdToken,
        [property: JsonPropertyName("refreshToken")] string RefreshToken,
        [property: JsonPropertyName("expiresIn")] string ExpiresIn,
        [property: JsonPropertyName("localId")] string LocalId,
        [property: JsonPropertyName("email")] string? Email,
        [property: JsonPropertyName("displayName")] string? DisplayName);
}
